examples/petsc/bpsdmplex.c

```diff
@@ -0,0 +1,752 @@
+//     bps -problem bp3 -ceed /gpu/occa -petscspace_degree 3
+//     bps -problem bp4 -ceed /cpu/occa -petscspace_degree 3
+//     bps -problem bp5 -ceed /omp/occa -petscspace_degree 3
+//     bps -problem bp6 -ceed /ocl/occa -petscspace_degree 3
+//
+//TESTARGS -ceed {ceed_resource} -test -problem bp3 -petscspace_degree 3
+
+/// @file
+/// CEED BPs example using PETSc with DMPlex
+/// See bps.c for a "raw" implementation using a structured grid.
+const char help[] = "Solve CEED BPs using PETSc with DMPlex\n";
+
+#include <stdbool.h>
+#include <string.h>
+#include "common.h"
+#include <petscdmplex.h>
+#include "bp1.h"
+#include "bp2.h"
+#include "bp3.h"
+#include "bp4.h"
+
+#define PATH(BASE) __DIR__ #BASE
+
+// Get CEED restriction data from DMPlex
+static int CreateRestrictionPlex(Ceed ceed, CeedInt P, CeedInt ncomp,
+				 CeedElemRestriction *Erestrict, DM dm) {
+  PetscInt ierr;
+  PetscInt c, cStart, cEnd, nelem, ndof, *erestrict, eoffset;
+  PetscSection section;
+  Vec Uloc;
+
+  PetscFunctionBegin;
+
+  // Get Nelem
+  ierr = DMGetDefaultSection(dm, &section); CHKERRQ(ierr);
+  ierr = DMPlexGetHeightStratum(dm, 0, &cStart,& cEnd); CHKERRQ(ierr);
+  nelem = cEnd - cStart;
+
+  // Get indices
+  ierr = PetscMalloc1(nelem*P*P*P, &erestrict); CHKERRQ(ierr);
+  for (c=cStart, eoffset=0; c<cEnd; c++) {
+    PetscInt numindices, *indices, i;
+    ierr = DMPlexGetClosureIndices(dm, section, section, c, &numindices,
+                                   &indices, NULL); CHKERRQ(ierr);
+    for (i=0; i<numindices; i+=ncomp) {
+      for (PetscInt j=0; j<ncomp; j++) {
+        if (indices[i+j] != indices[i] + copysign(j, indices[i]))
+          SETERRQ1(PETSC_COMM_SELF, PETSC_ERR_ARG_INCOMP,
+                   "Cell %D closure indices not interlaced", c);
+      }
+      // Essential boundary conditions are encoded as -(loc+1)
+      PetscInt loc = indices[i] >= 0 ? indices[i] : -(indices[i] + 1);
+      erestrict[eoffset++] = loc/ncomp;
+    }
+    ierr = DMPlexRestoreClosureIndices(dm, section, section, c, &numindices,
+                                       &indices, NULL); CHKERRQ(ierr);
+  }
+
+  // Setup CEED restriction
+  ierr = DMGetLocalVector(dm, &Uloc); CHKERRQ(ierr);
+  ierr = VecGetLocalSize(Uloc, &ndof); CHKERRQ(ierr);
+
+  ierr = DMRestoreLocalVector(dm, &Uloc); CHKERRQ(ierr);
+  CeedElemRestrictionCreate(ceed, nelem, P*P*P, ndof/ncomp, ncomp,
+                            CEED_MEM_HOST, CEED_COPY_VALUES, erestrict,
+                            Erestrict);
+  ierr = PetscFree(erestrict); CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+}
+
+// BC functions
+PetscErrorCode BCsMass(PetscInt dim, PetscReal time, const PetscReal x[],
+                       PetscInt vscale, PetscScalar *u, void *ctx) {
+  for (PetscInt i = 0; i < vscale; i++)
+    u[i] = PetscSqrtScalar(PetscSqr(x[0]) + PetscSqr(x[1]) +
+                           PetscSqr(x[2]));
+
+  return 0;
+}
+
+PetscErrorCode BCsDiff(PetscInt dim, PetscReal time, const PetscReal x[],
```

Comment:
@jedbrown, with single source QFunctions on the horizon, I did not use this function in the setup QFunctions like you did in your branch. It would cause problems for the code generation backends.

Reply:
Yeah, okay. That can be left for the future.